cmd/indexer/internal/datastore/pq/public_key.go

```diff
@@ -0,0 +1,100 @@
+package pq
+
+import (
+	"context"
+	"time"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+	"github.com/mailchain/mailchain/cmd/indexer/internal/datastore"
+	"github.com/mailchain/mailchain/crypto"
+	"github.com/mailchain/mailchain/crypto/multikey"
+	"github.com/pkg/errors"
+)
+
+// PublicKeyStore database connection object
+type PublicKeyStore struct {
+	db *sqlx.DB
+}
+
+// PublicKeyStore create new pointer to postgres database
+func NewPublicKeyStore(db *sqlx.DB) (datastore.PublicKeyStore, error) {
+	return &PublicKeyStore{db: db}, nil
+}
+
+type publicKey struct {
+	Address       []byte    `db:"address"`
+	PublicKey     []byte    `db:"public_key"`
+	CreatedAt     time.Time `db:"created_at"`
+	UpdatedAt     time.Time `db:"updated_at"`
+	Protocol      uint8     `db:"protocol"`
+	Network       uint8     `db:"network"`
+	PublicKeyType uint8     `db:"public_key_type"`
+}
+
+func (s PublicKeyStore) PutPublicKey(ctx context.Context, protocol, network string, address []byte, pubKey crypto.PublicKey) error {
+	p, n, err := getProtocolNetworkUint8(protocol, network)
+	if err != nil {
+		return errors.WithStack((err))
+	}
+
+	uPublicKeyType, err := getPublicKeyTypeUint8(pubKey.Kind())
+	if err != nil {
+		return errors.WithStack((err))
+	}
+
+	sql, args, err := squirrel.Update("public_keys").
+		Set("public_key_type", uPublicKeyType).
+		Set("public_key", pubKey.Bytes()).
+		Set("updated_at", time.Now()).
+		PlaceholderFormat(squirrel.Dollar).
+		Where(squirrel.Eq{"protocol": p}).
+		Where(squirrel.Eq{"network": n}).
+		Where(squirrel.Eq{"address": address}).
```

Comment:
`protocol`, `network`, `address` are part of the key and `public_key`, `public_key_type` is the value. I do not think the value is expected to change but it could be a possibility. Now I think about it more the time is not relevant, the block hash and tx hash is more useful information. What do you think to fields: 
- `created_block_hash`
- `updated_block_hash`
- `created_tx_hash`
- `updated_tx_hash`
- remove `created_at`
- remove `updated_at`

This would required altering `cmd/indexer/db/migrations/000001_public-key-index.up.sql` its ok to directly edit this file rather than create a new migration as the DB is not in production yet.

Reply:
Have modified the schema and the code to incorporate the above.

Have a few newbie questions here:
1) Should we also have `public_key_type` as part of the key? So a a particular address on a blockchain protocol/address can have multiple keys, one per key type? I believe every chain will generally have a single crypto algorithm behind it, so maybe I am wrong here
2)  When the public key changes on a blockchain, will the change be reflected on a block? I am a bit confused as to the relationship between public keys and the block/tx.

Reply:
Thanks @ramkumarvenkat 

Answering some of your good questions:
1. It is possible for a protocol to have multiple public key types / crypto algorithms [substrate supports 3 curve types](https://medium.com/polkadot-network/kusama-cc-2-prepos-a0ce785bb629) . Each address will have one public key + type combination, this is what is used to calculate the address.
2. I am not expecting any changes of public key against an address that would indicate a problem. Being able to track the change is useful. I have see that ethereum has used [different signers](https://github.com/ethereum/go-ethereum/blob/3bb6815fc1ade61e85b7039cf2e08a0871bb47e4/core/types/transaction_signing.go#L42-L53) as it has iterated. This though does not look like it affects the public key contents. I think a good solution is to validate that the public key can `Verify` the transaction
3. I am not expecting any change in public key for an address as attached to the blockchain protocols. This could change so want to be able to monitor it. We use the public key that is associate with an account but in the future we will allow users to provide alternative public keys too. Currently re-encrypting old messages is out of scope.

Reply:
Thanks @robdefeo , your explanation clarifies a lot of things for me!!